consul/resource_consul_configuration_entry.go

```diff
@@ -0,0 +1,176 @@
+			Name:     configName,
+			Protocol: d.Get("protocol").(string),
+		}
+	case consulConfigEntryProxyDefaults:
+		config = &consulapi.ProxyConfigEntry{
+			Kind:   consulConfigEntryProxyDefaults,
+			Name:   configName,
+			Config: d.Get("config").(map[string]interface{}),
+		}
+	default:
+		return fmt.Errorf("Config kind '%s' is not supported", configKind)
+	}
+
+	wOpts := &consulapi.WriteOptions{
+		Token: d.Get("token").(string),
+	}
+	if _, _, err := configEntries.Set(config, wOpts); err != nil {
+		return fmt.Errorf("Failed to set '%s' config entry: %#v", configName, err)
+	}
+
+	return resourceConsulConfigurationEntryRead(d, meta)
+}
+
+func resourceConsulConfigurationEntryRead(d *schema.ResourceData, meta interface{}) error {
+	configEntries := getClient(meta).ConfigEntries()
+	configKind := d.Get("kind").(string)
+	configName := d.Get("name").(string)
+
+	qOpts := &consulapi.QueryOptions{
+		Token: d.Get("token").(string),
+	}
+	configEntry, _, err := configEntries.Get(configKind, configName, qOpts)
+	if err != nil {
+		if strings.Contains(err.Error(), "Unexpected response code: 404") {
+			// The config entry has been removed
+			d.SetId("")
+			return nil
+		}
+		return fmt.Errorf("Failed to fetch '%s' config entry: %#v", configName, err)
+	}
+
+	switch configKind {
+	case consulConfigEntryProxyDefaults:
+		if err = d.Set("config", configEntry.(*consulapi.ProxyConfigEntry).Config); err != nil {
+			return fmt.Errorf("Failed to set 'config': %#v", err)
+		}
+	case consulConfigEntryServiceDefaults:
+		if err = d.Set("protocol", configEntry.(*consulapi.ServiceConfigEntry).Protocol); err != nil {
+			return fmt.Errorf("Failed to set 'protocol': %#v", err)
+		}
+	default:
+		return fmt.Errorf("Config kind '%s' is not supported", configKind)
+	}
+
+	d.SetId(fmt.Sprintf("%s-%s", configKind, configName))
+
+	return nil
+}
+
+func resourceConsulConfigurationEntryDelete(d *schema.ResourceData, meta interface{}) error {
+	configEntries := getClient(meta).ConfigEntries()
+	configKind := d.Get("kind").(string)
+	configName := d.Get("name").(string)
+
+	wOpts := &consulapi.WriteOptions{
+		Token: d.Get("token").(string),
+	}
+	if _, err := configEntries.Delete(configKind, configName, wOpts); err != nil {
+		return fmt.Errorf("Failed to delete '%s' config entry: %#v", configName, err)
+	}
+	d.SetId("")
+	return nil
+}
+
+func validateConfigEntry(d *schema.ResourceData) error {
+	configKind := d.Get("kind").(string)
+	configName := d.Get("name").(string)
+
+	switch configKind {
+	case consulConfigEntryProxyDefaults:
+		if configName != "global" {
+			return fmt.Errorf("'name' must be 'global' when 'kind' is 'proxy-defaults'")
+		}
+		if d.Get("protocol").(string) != "" {
+			return fmt.Errorf("'protocol' must not be set when 'kind' is 'proxy-defaults'")
+		}
+	case consulConfigEntryServiceDefaults:
+		if len(d.Get("config").(map[string]interface{})) != 0 {
+			return fmt.Errorf("'config' must not be set when 'kind' is 'service-defaults'")
+		}
```

Comment:
This validation is great but I guess if we want this to be generic it would have to go. I wonder if it makes the UX bad if we just accept anything and rely on Consul reporting errors when we attempt to write it on apply.

I can see how having earlier feedback is nicer I just worry that the validation here will end up being a burden to keep in sync with the equivalent in Consul - for example when we add namespaces there will be valid `proxy-defaults` configs other than `global` as you'll be able to set one per namespace but anyone with the old provider won't work.